Pattern: Spoken

antonjs opened the pull request:
“Do you not know that a man is not dead while his name is still spoken?”
         -- Terry Pratchett, Going Postal

Roughly inspired by the above quote, and the story in the book, this pattern takes a list of names of those we wish to remember, and pulses them out using morse code on a single lamp. Parameters are how many of the total temple lamps to use, the rate at which new names are started, pulsing speed, and speed variance. The idea is that people could submit names and have the temple speak them in morse throughout the week.

I can imagine two modes of use:

1. The full lamp set is used: people can watch knowing that each pulsing light is a name special to them or somebody else, beating into the night. The pulses are colored or overlaid with other effects (as with a normal pattern).
2. The pattern is run on a priority channel, but very sparsely. When a bright light starts pulsing on top of whatever else is playing, we know a name is being spoken.

The included Spoken.lxp demonstrates both of these modes.

The code works, but there are definitely some inefficient choices (although it seems to run fine at high framerate). Happy to fix / change things that need work, particularly since I'm still trying to come to grips with the API. I'd love any feedback.

If folks think this is an interesting idea then I can try to coordinate a way of collecting names special to people, to be pulsed through the pattern, both online in advance of the burn and perhaps on-playa if updating the dataset is at all feasible.

Cheers!

antonjs committed:
Slow the roll of the background pulses.

antonjs committed:
Spoken initial commit: pulse names out in morse code.

antonjs committed:
Merge branch 'master' of https://github.com/temple2018/Galaxia

antonjs committed:
Fix Terry's name. Sorry.

antonjs committed:
Fix gnarly resource leak; remove some debug prints.

antonjs committed:
Access Spoken names using relative path, update file name.

antonjs committed:
Run spotlessApply to match CI format reqs.

Geethree commented:
This is a neat idea. I'm about to take off will @mcslee or @deleted please review?

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+
+@LXCategory("Acjs")
+public class Spoken extends TemplePattern {
+  private static final String NAME_DATA = "/Users/anton/Projects/Galaxia/Galaxia/assets/names.txt";
```
Don't think you should use an absolute path here... Processing has helper methods that load files from the base directory, refer to this as "assets/names.txt" - you might also want to use a filename here that is more obviously related to the pattern file, e.g. Spoken.txt

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+              .setDescription("How variable dot length will be between names.");
+
+  private final HashMap<Character, String> morseLookup = new HashMap<>();
+  private final List<String> names = new ArrayList<>();
```
Nitpicky but those ArrayList<>() declarations should have the type inside the <>

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  void loadNames() {
+    Path dataPath = Paths.get(NAME_DATA);
```
See: https://processing.org/reference/loadStrings_.html

The absolute file path here isn't going to load for other people, relativize that and use the Processing helper to get a String[] in a single call.

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+        e.printStackTrace();
+      } finally {
+        // Backup / default
+        names.add("Larry Harvey");
```
Do you want two copies of Larry Harvey in the case that this succeeds? I guess that's fine but just checking.

antonjs replied:
Agree; the idea was just to have something if the absolute path load failed (didn't know about the Processing helpers, that's nice). Maybe it doesn't hurt to double his chances though...

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+    double dotLength = dot.getValue();
+    dotLength += (Math.random() - 0.5) * dotVariance.getValue();
+
+    Collections.shuffle(availableLights);
```
This doesn't really matter, but it would be a lot quicker to just generate a random int between [0, availableLights.size()-1]  and then grab that out of the list, rather than shuffling the entire list to get one random item.

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+
+    MorsePulser pulser = new MorsePulser(lx, light, message, dotLength);
+    pulsers.add(pulser);
+    addLayer(pulser);
```
It doesn't look like you ever call removeLayer() for these pulsers, even when they're finished. Over time the LXPattern will have a huge number of layers. These objects either need to be removed or re-used.

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+      MorsePulser pulse = iterator.next();
+      if (pulse.isFinished()) {
+        availableLights.add(pulse.getLightIndex());
+        iterator.remove();
```
I think you also want to removeLayer() here. Otherwise the layer still stays on the LXPattern and its run() method will be called every frame.

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+  public void run(double deltaMs) {
+    // Give each unlit 'slot' a chance to be born.
+    for (int i = 0; i < (int) density.getValue() - pulsers.size(); i++) {
+      if (names.size() == 0) loadNames(); // Reload names if we're out. XXX: Performance?
```
Yeah, this definitely feels off. Reloading the names is going to give you the same list every time (just shuffled differently). I think it makes more sense to just shuffle it once on initial load and then loop through them. Or you can periodically shuffle again if you don't like the idea of it being in the same loop-order.

antonjs replied:
The idea here was that we could update the underlying file during run with new names, captured through some kind of TBD on-playa means, and that they would get loaded in. It definitely doesn't need to be reloaded every time the list is exhausted though.

If you could give me an idea of the lighting setup on-playa (a headless machine running the same LX instance all burn versus different laptops/instances being switched in and out, for instance), happy to come up with a better solution to hot reloading. If you don't think it's practical to update at all, I can take it out.

deleted replied:
@antonjs It's going to be on a headless linux machine in a NEMA enclosure at the base of the temple, and I believe we'll also have a wireless access point on it. If you want to try and devise some means of getting user input, we could probably set you up with an account on that box with permissions to transfer files to a specific place on that box.

Let us know if you work out more specific plans about how to collect names, but be advised that over the next few days a lot of us will be to playa and might be harder to reach after that.

antonjs replied:
Thanks @deleted. Current thinking is an retro terminal that would allow user input, located somewhere that works for you all / the broader temple crew. Uploading a text file over wifi would definitely work. I can put together a more detailed plan and get it to you asap.

I'll be on playa from 22 Aug (Wednesday) so can coordinate in person then, assuming of course you don't have more urgent chaos to attend to...

deleted replied:
That sounds very cool. Keep us posted!

I've noticed that @alanperpersak has also submitted a pattern that [encodes and displays string messages](https://github.com/temple2018/Galaxia/pull/32/files#diff-dbd65fe961ae37d61153c255d1e0f413). Perhaps the two of you would like to coordinate on input sources. I'll drop a note in his PR as well.

antonjs replied:
Thanks @deleted, I'll drop Alan a line. I've added a (slightly) more fleshed out proposal for doing the collecting names below, please let me know what you think!

mcslee commented on src/main/java/org/templegalaxia/patterns/acjs/Spoken.java:
```diff
@@ -0,0 +1,216 @@
+          return;
+        }
+
+        currentModulator().start();
```
I think this is fine, but currentModulator().trigger() may be more semantically appropriate (start means resume running from wherever it was, trigger() means start a new cycle from the beginning).

antonjs replied:
Great to know. I feel like I tried trigger first and it wasn't working, but that was likely while something else was broken.

mcslee reviewed (changes requested):
This looks fine in general. Left comments on various things that I'd clean up.

The only major one that really matters is the absolute file path pointing at "/Users/anton" - which will break on anyone else's machine. So just get that one cleaned up and use the Processing helper functions to load the file, then we're good to go. Can clean up the other issues in here at your own discretion. The missing removeLayer() is the other critical one, as that will currently stack up a massive number of layers on the pattern.

mcslee commented:
Oh, and it's a very lovely idea by the way! Really neat way of using these lights in line with what the Temple is for.

antonjs commented:
Thanks for all the feedback @mcslee. I'll update the pull when I can.

If you all are happy to go ahead with this, I'll start on a way for people to submit names. Given that it is symbolic and that I'll need to advertise to the community, is there someone I should connect with on the Temple project around the idea and messaging? Or just go do it?

Cheers!

mcslee commented:
That's really a question for Gerald... I'm not really an official Temple
team member and have just been helping with the software. I know the topic
of animated light on the Temple has already had some sensitivity around it,
Gerald might be able to better opine on a good process for how this is
incorporated, or tell you who the right person to ask would be.

I do think that this concept will really resonate with people though - pays
such great to homage to what the Temple means to so many folks.

On Wed, Aug 8, 2018 at 3:42 PM Anton Jackson-Smith <notifications@github.com>
wrote:

> Thanks for all the feedback @mcslee <https://github.com/mcslee>. I'll
> update the pull when I can.
>
> If you all are happy to go ahead with this, I'll start on a way for people
> to submit names. Given that it is symbolic and that I'll need to advertise
> to the community, is there someone I should connect with on the Temple
> project around the idea and messaging? Or just go do it?
>
> Cheers!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/temple2018/Galaxia/pull/29#issuecomment-411576825>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkfWlcFIsHKIucjZn5q1TR-RN_PIcn-ks5uO2lagaJpZM4VvmKV>
> .
>

antonjs committed:
Cleanup ArrayLists and select random light more efficiently.

This is why we don't code while exhausted.

antonjs committed:
Remove pulse layer when it finishes.

antonjs committed:
Start Spoken modulators using trigger()

antonjs committed:
Remove unused imports; Spotless

antonjs committed:
Merge branch 'master' into master

antonjs committed:
Update name list every 10 minutes at most.

antonjs committed:
Merge remote-tracking branch 'origin/master'

antonjs commented:
@mcslee Thanks for the kind words on the concept. I think I've addressed all of your feedback in the subsequent commits; let me know if there's anything else that needs changing.

@deleted @Geethree, I've attached a brief proposal of the overall project and our methods for on- and off-playa name collection to this comment. If you could take a look and/or get them in front of whoever is most appropriate to give the thumbs up that would be amazing. Ideally we'd be able to send out the website to collect names this weekend.

Happy to answer any questions you might have, cheers!

[Spoken Proposal.pdf](https://github.com/temple2018/Galaxia/files/2280635/Spoken.Proposal.pdf)

Geethree commented:
Hi hi hi! 

Internet is a little spotty out here. I’ll be frank, I enjoy this idea but I’m a little weary of the on playa component. Granted, if it’s self contained / plug and play I see no reason why to say no. 

For the pre playa, let’s do it. What do you need from me?


For the on playa, when do you all arrive? I’m afraid the temple is full up and cannot offer any WAPs. 

Let me know. 

As per merging, @mcslee if you can give it another once over. I trust you to merge. :)

antonjs commented:
Cheers @Geethree. We'll give the execution of on-playa our best shot and see what we can make happen by the time we get out there. We're both coming out early (Wed 22) to build DeMaTerial at 7:45/I, so good on the WAP front.

Pre-playa, we don't need anything, although if you wanted to signal boost our messages about the website (http://galaxiaspoken.org/) through the Temple's channels that would be amazing.

On-playa, we just need access to the wireless network connected to the LX system, and a sense of how we should plan to transfer files to the lighting machine (SCP / FTP / some REST request / floppy disks / whatever). Once we're arrived and situated I can come and attempt to track you down (no idea how that could go wrong :P).

Hope things are going well out in the desert!

antonjs committed:
Merge branch 'master' of https://github.com/temple2018/Galaxia

antonjs committed:
Merge branch 'master' into master

antonjs commented:
Friendly bump, @mcslee

mcslee commented:
Sorry about that - sorted now!

antonjs commented:
No problem. We’ve managed to collect >2700 names from people online, so excited to see this in action :)

> On Aug 16, 2018, at 1:28 PM, Mark Slee <notifications@github.com> wrote:
> 
> Sorry about that - sorted now!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/temple2018/Galaxia/pull/29#issuecomment-413674280>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ADHyyruEg4st1sm9Bja6MioBr_zApzbuks5uRdXWgaJpZM4VvmKV>.
>
